[#46555] Extend Query resource in APIv3 by timestamps

dombesz opened the pull request:
https://community.openproject.org/work_packages/46555

- [X] POST /api/v3/queries (create queries) accepts a timestamp parameter in the message body
- [X] PATCH /api/v3/queries/:id (update query) accepts a timestamp parameter in the message body
- [X] POST /api/v3/queries/form and POST /api/v3/queries/:id/form (form for creating and updating queries) accept timestamp parameters in the message body and the embedded schema explains the added parameter
- [X] ~~GET /api/v3/projects/:id/queries/default, /api/v3/queries/default and /api/v3/queries/:id accept the additional url parameter 'timestamps' similar to how it is done for work packages.~~ Seems to already be supported. Double checking required. Added tests to ensure.
 - [X] API documentation is updated.

dombesz committed:
[#46555] Extend Query resource in APIv3 by timestamps

https://community.openproject.org/work_packages/46555

ulferts commented on app/models/query/timestamps.rb:
```diff
@@ -32,6 +32,9 @@ module Query::Timestamps
   included do
     serialize :timestamps, Array
 
+    validates :timestamps, length: { maximum: 2 }
```
Is there a necessity to limit it to 2?

ulferts commented on app/models/query/timestamps.rb:
```diff
@@ -57,5 +60,15 @@ def timestamps=(array)
+      if invalid_timestamps.any?
+        errors.add :timestamps, :invalid, values: invalid_timestamps.join(", ")
+      end
+    end
```
The default for validations in the OP architecture is to have them in the contracts. The contracts will then be the place to check for a model's validity. This has the benefit of having different contracts for different situations (e.g. create vs update). That is not the case throughout the application, yet but we should work towards that.

ulferts commented on lib/api/v3/queries/schemas/query_schema_representer.rb:
```diff
@@ -122,6 +122,12 @@ def self.filters_schema
                  deprecated: true
 
+          schema :timestamps,
+                 type: '[]Timestamp',
```
The type `Timestamp` is current not clearly defined in the API documentation. Adding and entry to the [section on Date, Time and Duration objects](https://www.openproject.org/docs/api/basic-objects/#dates-times-and-durations) would help.

ulferts reviewed (changes requested):
Hi @dombesz ,

I found a couple of things but overall, the change is looking good. 

The `timestamps` property is also part of the schema by this PR already so it is returned when fetching `GET http://localhost:3000/api/v3/queries/schema`. 

## Error on faulty timestamp declaration

Posting to the query form (`POST http://localhost:3000/api/v3/queries/form`) with a body of: 
```
{
   "timestamps": "PT0S"
}
```
(providing only a single value instead of an array with one item) causes an internal error

## Faulty timestamp saved to database

Creating a query (`POST http://localhost:3000/api/v3/queries`) with a body of: 
```
{
    "name": "timestamp query",
   "timestamps": ["2022-03-02 invalid string 20:45:56Z", "PT0S"]
}
```
stores that invalid timestamp value in the database. Everything seems to be stable afterwards but it would be better if we would only store a format that is so8601 compatible. 

## Documentation

The timestamp property needs to be added to the query resource documentation for the API v3.

dombesz committed:
fixup! [#46555] Extend Query resource in APIv3 by timestamps

dombesz committed:
fixup! [#46555] Extend Query resource in APIv3 by timestamps

dombesz committed:
fixup! [#46555] Extend Query resource in APIv3 by timestamps

dombesz committed:
fixup! [#46555] Extend Query resource in APIv3 by timestamps

dombesz committed:
Fix Timestamp class rubocop error.

dombesz commented:
Thanks @ulferts for the review! I addressed all your comments.
There is one additional change I made regarding your comment on the "[Faulty timestamp saved to database](https://github.com/opf/openproject/pull/12357#pullrequestreview-1372474774)" .

The `Timestamp.parse` method was using `Time.parse` which was very forgiving and it validated the problematic string (`"2022-03-02 invalid string 20:45:56Z"`) correctly. A more strict parsing method is now applied, it accepts the duration format ("PT0S") and the ISO8601 datetime format. Everything else throws a validation error. Please let me know if this solution is desirable, or if we need to look for a different one.

cbliard committed:
Add requestBody to some queries API endpoints

This allows tools relying on OpenAPI like restish to use these endpoints properly.

cbliard committed:
Add timestamps parameter to OpenAPI GET query endpoints

Also rearrange fields order for better readibility in source code (name first).

cbliard reviewed (approved):
It looks good and works as expected as far as I have tested it :+1: 

Some minor remarks which can improve the documentation and should not prevent merging:

- documentation of `timestamps` property is missing from the Query model schema (in `docs/api/apiv3/components/schemas/query_model.yml`).

- I was wondering where it is documented in the OpenAPI that the `Timestamp` type can be either a ISO8601 duration, a ISO8601 date (are timestamps allowed?), or one of the special values `1y`, `1m`, `1w`, `1d` and their combinations / negatives? I finally I found it: it's in `docs/api/apiv3/tags/basic_objects.yml`. Maybe the paragraph title `Dates, Times and Durations` should be renamed as `Dates, Times, Durations, and Timestamps`. For completeness, it could also mention the special values (`1y-2m` for instance)

- The examples in the API documentation are missing the timestamps field for the GET query/{id} endpoint (having `timestamps: ["PT0S"]` in the response example would probably be enough)

- Should the documentation mention that the timestamps attribute is present only on Enterprise edition (for instance `highlightedAttributes` linked property has the bold text **Only with valid Enterprise Token available**) or is it part of the things to be done in another work package?

dombesz commented:
Thanks @cbliard, added modifications on the documentation [here](https://github.com/opf/openproject/pull/12416/commits/f8e51a3a1190c9f916409f155dbde3b9adfc2370).
